Goofs Old Monk PR(#35535) reworked a little to work with 12/19 codebase. Don't try revive until a decent martial art rework.

davethwave opened the pull request:
It's the return of goof's last year monk PR. I am a newbie in Dream Maker code, I am sure I made a few mistakes when I updated his PR to make it work with the recent changes that have happened so make sure to point out any mistakes so I can fix them. As of right now it works as the old monk did previously except that it's ONLY admeme accessible so no more chaplain monk. While it was fun to be chaplain monk it is too powerful as is. In the future if this is merged and the antag freeze is over I'll think of making this a chaplain traitor weapon for so and so tc or a new chaplain null rod with a very restricted max level like 9 or something. @Iamgoofball https://github.com/tgstation/tgstation/pull/35535

:cl:
add: Added in Goof's old monk martial art that only admins can grant to people.
/:cl:

davethwave committed:
Goofball's monk

The monk is back!

davethwave committed:
No Chaplain monk

Chap monk be too OP

davethwave committed:
Quick fix

oops

81Denton commented:
>It's the return of goof's last year monk PR, as I am a newbie in Dream Maker code

oh no no no

davethwave committed:
freaking trailing newlines

Iamgoofball commented:
@davethwave the boys are back in town

this should, instead, be a lavaland ruin hidden inside of a temple with kung fu zombies who you have to fight first

davethwave commented:
Maybe after it's merged, it's kinda OP as is so I think it should be a admin only thing. Monk traveling to station to impart lost art or admeme event for now, and or give chaplain a very restricted version of it. Would be helpful if it can be looked at for any potential problems though. If it makes you feel better you can get the disciple books from Christmas presents if this gets testmerged/merged while the Christmas tree thing is happening.

dr-bee commented:
I had a blast testing this when goof originally worked on it. Favorite bug was roundhouse kick hitting everyone including the monk.

It was unbalanced as hell but fun. with a few tweaks it could be made a useful addition, such as having it only deal stamina damage, and giving a version of pacifism.

ATH1909 commented:
what it is what it ain't what it isn't

Shadowflame909 commented:
Return of the Monk. (Yes it is, Yes it iiiuhhh!)

is it an admin variant of a nullrod or just under the martial arts VV

davethwave commented:
You have to VV it. It's not like it's super difficult so yeah. Also you can also VV to make someone a apprentice so they can't go past level 19 and give other people monk powers.

davethwave commented:
Also on a side note, the monk has been slightly nerfed as somehow some viro code was breaking it. Originally you would be virus immune due to monk power but now monk have normal human immune system.

necromanceranne commented:
That's because 3.5 DnD monks were immune to disease. It was the Purity of Body power you get at level 5.

I'll need to check the code but goof mentioned Diamond Body being a power as well. That was poison immunity so I guess you're immune to toxins?

Iamgoofball commented:
@davethwave Yeah, it wasn't viro code breaking, it was intentional. Monks can't get disease. Purity of Body.

Iamgoofball commented:
@necromanceranne ya, every 3.5e monk ability was in

necromanceranne commented:
There are a few of those that are straight up wizard spells. Abundant Step is Dimension Door/Blink on tg, Empty Body is Etherealness/Ethereal Jaunt on tg, sorta Quivering Palm is to tg's Smite/Disintergrate even if that isn't as much a violent death. I have to check again but I do know if you added those.

But damn dude nice work translating over the class into the game that's really cool.

Shadowflame909 commented:
Monk Class is pretty lit, but when does it get the ability to have its spirit guarding itself.

Stando reference

necromanceranne commented:
Keep your jojo memes out of the DnD reference.

That's Pathfinder. And also the Summoner class. And Spiritualist actually. Mostly the latter.

Shadowflame909 commented:
Ah... Now I know the class I must get.


Dave the next PR you should revive is Goof's Vampire Mask PR. Let the item appear from the necropolis tendril, changing ones race from human to vampire.

optimumtact commented:
I'm very much not a fan of martial arts code

:-1:

nemvar commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+	button_icon_state = "neckchop" // todo: replace
+	uses_left = 4
+	max_uses = 4
+	skill_name = "stunning_fist"
```
Link this to type. Magic strings big gay.

davethwave replied:
Will try to figure out how.

davethwave replied:
Think this is the only meaningful issue left I believe, I'll see if I can get some help or figure it out on my own.

davethwave replied:
I believe I did what you wanted, however I'm not sure. Put another comment here if this isn't.

nemvar commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+/datum/martial_art/monk/process()
+	..()
+	if(diamond_body)
+		owner.reagents.remove_all_type(/datum/reagent/toxin, 30)
```
Just give the TOXIMMUNE trait.

davethwave replied:
I'll take care of it soon and add a anti virus trait the name I forgot of too.

Iamgoofball replied:
Toximmune didn't exist at the time of this coding, my apologies.

davethwave commented:
By viro code breaking it I mean that it wouldn't compile.

Iamgoofball commented:
@optimumtact Is there a newer code method we could use to accomplish this? At the time of the coding of Monk martial arts was the "cleanest" way to go about a system like this.

Iamgoofball commented:
also you should give it back to chaplains tbh

optimumtact commented:
there isn't a newer method, martial arts are just shit

Iamgoofball commented:
ya, it sucks a lot and I had to do some dumb workarounds for the monk abilities on-punch to work with martial arts code, but at the same time i think the ei nath method of making a hand object to do it is equally snowflakey

Iamgoofball commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+			to_chat(owner, "<span class = 'danger'>You have practiced the Quivering Palm technique enough to use it in combat.</span>")
+			quivering_palm.Grant(owner)
+			quivering_palm.martial = src
+			available_actions = list(stunning_fist, wholeness_of_body, quivering_palm)
```
update the list instead of making a brand new one, this one's my bad

davethwave replied:
Done and done.

Iamgoofball commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+	if(atr)
+		var/old_stat = D.stat
+		if(D.stat != DEAD && D.ckey && D != A)
+			add_exp(5)
```
please make sure this check is updated with any additional "make sure this person is actually a player" changes made since the original PR

davethwave replied:
Believe everything here is still in working order from what I can tell.

Iamgoofball commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+				switch(use_cleave)
+					if(1)
+						do_attack(A, pick(surrounding_mobs), FALSE, FALSE, null)
+					if(2)
```
swap these magic numbers for defines

davethwave replied:
Did some research these few days and don't quite know how to do what you want unfortunately. I'll see if I can contact some people or ask from help to see if there's a solution.

davethwave replied:
Thanks to some outside help I've found a idiotically simple answer to the problem

Iamgoofball commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+		if(H.mind && istype(H.mind.martial_art, /datum/martial_art/monk))
+			var/datum/martial_art/monk/M = H.mind.martial_art
+			return M.defense_roll(curr_ab + abm)
+	var/armor_class = 10 + (T.getarmor("chest", "melee") * 0.1)
```
Something I forgot to do here was factor in the monk's armor class and use that to give them a penalty for heavy armor, but we might not be able to actually do that within the context of SS13

davethwave replied:
So take it out entirely?

Iamgoofball replied:
What? No

davethwave replied:
So what do you want me to do exactly?

Iamgoofball commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+	var/armor_class = 10 + (T.getarmor("chest", "melee") * 0.1)
+	var/attack_bonus = curr_ab + abm
+	if(T.incapacitated())
+		attack_bonus += 3
```
May want to swap this to using proper Coup De Grace rules, but that might be a little extreme.

> A full-round action that allows an attacker to attempt a killing blow against a helpless opponent. A coup de grace can be administered with a melee weapon, or with a bow or crossbow if the attacker is adjacent to the opponent. An attacker delivering a coup de grace automatically scores a critical hit, after which the defender must make a Fortitude save (DC 10 + damage dealt) or die. Rogues also gain their extra sneak attack damage for this attack. Delivering a coup de grace provokes attacks of opportunity from threatening foes. A coup de grace is not possible against a creature immune to critical hits.

necromanceranne replied:
I mean if you are in a state of complete helplessness and the attack takes a reasonable amount of time, nothing really changes here. You could just as easilly kill them with kicks. Sleeping Carp has Elbow Drop, which is basically a guaranteed Coup De Grace.

davethwave replied:
Reminder but if I remember right a full *round* in dnd is 3 seconds.

necromanceranne replied:
6 seconds.

Iamgoofball commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+		armor_class -= armor
+	var/attack_bonus = abm
+	if(owner.incapacitated())
+		attack_bonus += 3
```
see above comment about Coup De Grace

optimumtact commented on code/_onclick/item_attack.dm:
```diff
@@ -56,7 +56,11 @@
+		if(istype(user, /mob/living/carbon/human))
+			var/mob/living/carbon/human/H = user
+			if(istype(H.mind.martial_art, /datum/martial_art/monk))
+				to_chat(user, "<span class='warning'>Your vows as a Monk prevent you from using [src] in combat!</span>")
```
make a trait

davethwave replied:
Will do

davethwave replied:
Think I did what you wanted but I'm not sure.

davethwave replied:
Did some testing with the original stuff and this bit of code isn't working right now even before I messed with it but was on the original 2018 version. Will see if I can fix.

davethwave replied:
Well I found the issue that broke it, apparently I didn't port this correctly at this part of the code but it works now. I however after fixing it have tried tying this with the monk trait however no matter what I did as soon as I added a 	if(HAS_TRAIT(H, MONK_TRAIT)) it would break again and wouldn't work. Unless you or someone else can figure out a way to get it to work I will consider this resolved.

optimumtact commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1451,6 +1451,29 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	armor_block = min(90,armor_block) //cap damage reduction at 90%
 	var/Iforce = I.force //to avoid runtimes on the forcesay checks at the bottom. Some items might delete themselves if you drop them. (stunning yourself, ninja swords)
 
+	if(H.mind && H.mind.martial_art && istype(H.mind.martial_art, /datum/martial_art/monk))
```
this is an awful bit of code and reveals everything that needs to be said about the martial arts system.

It's not self contained, it gets all through other code and snowflakes it up.

This is really bad

davethwave committed:
monk now uses virus/toximmune traits

did the thing

Iamgoofball commented on code/datums/martial/monk.dm:
```diff
@@ -0,0 +1,455 @@
+/datum/martial_art/monk/process(mob/living/carbon/human/H)
+	..()
+	if(diamond_body)
+		ADD_TRAIT(H, TRAIT_VIRUSIMMUNE, TRAIT_TOXIMMUNE)
```
incorrect
add TOXIMMUNE at level 11's level up, and add VIRUSIMMUNE at level 5's level up

davethwave replied:
Ok

davethwave replied:
Should be fixed now

davethwave commented:
Welp, discovered new bug. The monk is impossible to kill via melee right now. Even if blood is shown the person who has monk isn't hurt but can be hurt via thrown items. If someone finds out why and shows where before I do it would be greatly appreciated.

davethwave committed:
Think I did this right

Turns thing optimumtact into part of the monk trait

davethwave commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1451,6 +1451,29 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	armor_block = min(90,armor_block) //cap damage reduction at 90%
 	var/Iforce = I.force //to avoid runtimes on the forcesay checks at the bottom. Some items might delete themselves if you drop them. (stunning yourself, ninja swords)
 
+	if(H.mind && H.mind.martial_art && istype(H.mind.martial_art, /datum/martial_art/monk))
```
If my pitiful understanding of things is right the problem lays here somewhere for the no damage melee thing I mentioned.

davethwave replied:
Well farther down below where monk code I mean.

davethwave replied:
There was a chunk of code I also couldn't add in at code/datums/martial.dm fromhttps://github.com/tgstation/tgstation/pull/35535 due to a cleanup of martial code at https://github.com/tgstation/tgstation/pull/46497 that could also been possible.

davethwave replied:
Well I was right and with the help of the tg discord discord section the issue was found and fixed. Thanks 4dplanner.

davethwave committed:
Virus/Tox immune traits set to proper levels now(I think)

ATH1909 commented:
Guys, what if we used the skills system for this instead of the martial arts system?

PKPenguin321 commented:
It should use the skills system yes, but overriding unarmed attacks basically has to use martial arts as it's the only system for it

Iamgoofball commented:
we have a skills system now?

Tlaltecuhtli commented:
cant the skill system return attack combos?

GuyonBroadway commented:
Honestly a universal skills framework would probably be worth some effort.

davethwave commented:
I was thinking of trying to use the skill framework but it only has 9 levels or so if I remember.

davethwave commented:
> we have a skills system now?

It was added about a month and a half ago.
https://github.com/tgstation/tgstation/pull/46913

necromanceranne commented:
Skill system doesn't use a numerical level system it uses tiers of mastery. I'm pretty sure it only goes up to 5 tiers where as this is 1-30 levels and I don't think you can add tiers but I suppose you'd have to ask qust.

Character class and level system time.

davethwave committed:
Monk can now be esworded to death

gave a h to something and it fixes stuffs

davethwave committed:
Monk people can no longer use esword and stuff

davethwave committed:
updating lists

davethwave commented:
With the major bugs and problems out of the way I think that it can be test merged. Just need to work on the minor stuff like replacing magic numbers and stuff. It's at least as functional as it used be now.

davethwave commented:
By the way, for test merging purposes if and when we do so can I give the chaplain a null rod option with a monk scroll with a level limit of 9 or so?

MimicFaux commented:
> @davethwave the boys are back in town
> 
> this should, instead, be a lavaland ruin hidden inside of a temple with kung fu zombies who you have to fight first

Space ruin might be more appropriate, have it like a dojo with references to the space ninja.

Exavere commented:
Speaking from a slightly biased stand point here, but.  Even if it never gets merged as a real thing for players it's always fun having more Admin Tools to experiment with to spice things up even if it'll probably be used for Meme Smites on things like WGW/Smut posters.  So I'd like it even if it's Admeme only.

davethwave committed:
You now bald the more powerful you get

spookydonut commented on code/_onclick/item_attack.dm:
```diff
@@ -45,18 +45,21 @@
+		if(HAS_TRAIT(user, TRAIT_PACIFISM))
+			to_chat(user, "<span class='warning'>You don't want to harm other living beings!</span>")
+			return
+		if(istype(user, /mob/living/carbon/human))
```
```suggestion
		if(HAS_TRAIT(user, MONK_TRAIT))
			to_chat(user, "<span class='warning'>Your vows as a Monk prevent you from using [src] in combat!</span>")
			return
	
```

davethwave replied:
Thanks, when I have time I'll make the suggested change.

davethwave replied:
Tested out the code and unfortunately it doesn't work.

GuyonBroadway commented:
Having coded this for hippie where it lived for all of a week I discovered that if a monk manages to get gloves of the north star he can fight a room full of armed men and reliably win.

It is equal parts hilarious and terrifying.

necromanceranne commented:
Casting Haste on the monk was always the winning move.

davethwave commented:
Don't think it will be a problem as of right now as a admin only thing but if it does I'll see if there's a way to make it so that it doesn't work with the monk.

necromanceranne commented:
That'd probably involve making the gloves a martial art of their own, in which case it would actually hit all the other martial arts with it and revert it back to meme status.

Maybe not the worst move if it was a good martial art in its own right.

davethwave commented:
If I remember right hulk doesn't work well with the gloves so maybe I can research into why with a bit of code diving and find out a way to implement it here?

davethwave commented:
Also, thinking about it I could perhaps try to just restrict glove use in general like how you can't use headgear as a monk.

GuyonBroadway commented:
>That'd probably involve making the gloves a martial art of their own, in which case it would actually hit all the other martial arts with it and revert it back to meme status.

It's because the gloves rely on a standard unarmed attack using harm intent which hulk overrides and strangely martial arts don't.

davethwave committed:
Replaces some magic numbers

Thanks to Coul on Discord

davethwave committed:
Think this is what nemvar wanted

magic string begone(I think)

davethwave commented:
I believe that's all of them. All the reviews that I can do about have been taken care of.

davethwave committed:
Nerfs disciple

Max level is now 12

Iamgoofball commented:
give it back to Chaplains for the test merge @davethwave

davethwave commented:
Will do, want chap to get full monk power or just disciple?

davethwave committed:
monk null re-added for test but will be removed once ready for merge

Iamgoofball commented:
Chaplain gets full monk power. It created fun dynamics where people would come to him begging to be disciples and it was the funniest thing ever. And leave disciples at 20, it'll be fine IMO.

davethwave commented:
Reminder that their dodging ability also work on ranged weaponry.

Iamgoofball commented:
What? It's not supposed to. That's a bug.

TheChosenEvilOne commented on code/__DEFINES/maths.dm:
```diff
@@ -96,6 +96,9 @@
 // Returns the nth root of x.
 #define ROOT(n, x) ((x) ** (1 / (n)))
 
+// Real modulus that handles decimals
```
???

davethwave replied:
Ask Goof about that. It was in the original PR.

TheChosenEvilOne reviewed (commented):
Wha

davethwave commented:
It's a bug? I remember that being a ability back in the original but maybe I'm wrong. I will test it when I have time.

davethwave commented:
Did a bit and testing and it was a mistake of memory. No dodging of projectiles happen.

davethwave committed:
Disciple level max back to 20

davethwave committed:
Removes Null rod monk manual

Lack of maintainer interest and time the chaplain monk null rod has been open makes me think this won't get test merged

ExcessiveUseOfCobblestone commented:
Maintainer interest?

davethwave commented:
Welp, guess I will put the PR to sleep until a martial art rework happens. Maybe once martial code is in a better spot this will in turn also have a chance to be introduced and hopefully merged.

wesoda25 commented:
shame

Iamgoofball commented:
you're supposed to yell at headmins to do testmerges

davethwave commented:
I'll give this one more chance I guess.

Tharcoonvagh commented:
Make this a 20tc chaplain only traitor item

davethwave commented:
Antag freeze

GuyonBroadway commented:
Plus it's much nicer for the crew to have something to do.

And indeed in order to train properly the monk needs a willing participant who they can trust and trusts them to let them beat the shit out of them for a good 20 mins.

Plus if you give a disciple scroll to an antag and they buy gloves of the north star they now have the means to literally fistfight the whole station. They could go 4v1 in melee combat and win or instantly yeet a single person off this mortal coil in two seconds flat.

GuyonBroadway commented:
Forgot to add how in encourages cooperation but presents risks and requires trust so it could add an interesting dynamic.

necromanceranne commented:
I mean the really op aspects of it can be worked on later I'm sure. Daveth is already looking to prevent GotNS

davethwave commented:
Sorry if it's obvious or something but the heck is gotNS?

davethwave commented:
Really, it can be simply the full version is only admin grant only and other people can get a level limited version of it where it isn't so strong as a 2 punch knockout. Also armor is effective at weakening monk attacks by the way.

GuyonBroadway commented:
GotNS is Gloves of the North Star

davethwave commented:
On that note I believe that line 25 or 39,  of https://github.com/tgstation/tgstation/blob/master/code/datums/mutations/hulk.dm is whats bypassing the North Star gloves.

ExcessiveUseOfCobblestone commented:
https://github.com/tgstation/tgstation/pull/47986#issuecomment-563489888

davethwave commented:
A note for the future if this ever gets worked on again, there were a few runtimes I've noticed and a few icons that have to be replaced. There are a few choices I've found that can replace stuff like the healing button for the monk can be replaced with the fleshmend action button since it's just kind of a plus symbol that has a healing feeling about it. The heart attack punch thing can be replaced with the sniper zoom thing/ target thing since it's kinda like a critical hit kinda thing. The rest action button though I couldn't find a proper one already in the code can probably be made simply enough. A few simple Z's will suffice to get the point across and the flurry of fists can simply be a closed fist with it midair like it's about to punch someone or something and when you toggle it just add a X across it or something. The stunning fist can probably be replaced pretty simply though by just adding some lightning bolt effects to the fist you would use for the flurry of fists action button.
 The runtimes in question is about the lists for the abilities for when their added and the traits. The runtimes problems are related to these areas of code. Line 88 for the traits, line 115, and line 141. Both 115 and 141 are runtiming due to how how I tried to refine the code or however you put it. I attempted to turn it from making a new list every time you get new abilities like the stunning punch and the healing thing to just a updating list but it is runtiming for reasons I am not quite so sure on.
These are the main problems that I ran into. 
Also one last thing. Probably want to wait until martial arts gets reworked again before working on this or you won't get a chance. They want to avoid a clock cult situation where the thing you adds dips into multiple files in order to get the thing you want to work done.

Iamgoofball commented:
what is maintainer interest and does this mean i have to appease people that aren't the supreme overlord oranges?
